my/core/query.py

```diff
@@ -0,0 +1,386 @@
+) -> Optional[OrderFunc]:
+    """
+    Accepts an object Res[T] (Instance of some class or Exception)
+
+    If its an error, the generated function returns None
+
+    Most of the time, you'd want to provide at least a 'key', a 'where_function' or a 'default'.
+    You can provide both a 'where_function' and a default, or a 'key' and a default,
+    incase the 'where_function' doesn't work for a particular type/you hit an error
+
+    If a 'default' is provided, it is used for Exceptions and if an
+    OrderFunc function could not be determined for this type
+
+    If a key is given (the user specified which attribute), the function
+    returns that key from the object
+    tries to find that key on the object
+
+    Attempts to find an attribute which matches the 'where_function' on the object,
+    using some getattr/dict checks. Returns a function which when called with
+    this object returns the value to order by
+    """
+    if isinstance(obj_res, Exception):
+        if default is not None:
+            return lambda _o: default
+        else:
+            low(f"""While creating order_by function, encountered exception {obj_res}
+Value to order_by unknown, provide a 'default', filter exceptons with a 'where' predicate or
+pass 'drop_errors' to ignore this""")
+            return lambda _o: None
+
+    # shouldn't raise an error, as we return above if its an exception
+    obj: T = unwrap(obj_res)
+
+    if key is not None:
+
+        # in these cases, if your key existed on the initial Res[E] (instance that was passed to
+        # _generate_order_by_func and generates the OrderFunc)
+        # to run, but doesn't on others, it will return None in those cases
+        # If the interface to your ADT is not standard or very sparse, its better
+        # that you manually write an OrderFunc which
+        # handles the edge cases, or provide a default
+        # See tests for an example
+        # TODO: write test
+        if isinstance(obj, dict):
+            if key in obj:  # acts as predicate instead of where_function
+                return lambda o: o.get(key, default)  # type: ignore[union-attr]
+        else:
+            if hasattr(obj, key):
+                return lambda o: getattr(o, key, default)  # type: ignore[arg-type]
+
+    # Note: if the attribute you're ordering by is an Optional type,
+    # and on some objects it'll return None, the getattr(o, field_name, default) won't
+    # use the default, since it finds the attribute (it just happens to be set to None)
+    # should this do something like: 'lambda o: getattr(o, k, default) or default'
+    # that would fix the case, but is additional work. Perhaps the user should instead
+    # write a 'where' function, to check for that 'isinstance' on an Optional field,
+    # and not include those objects in the src iterable
+
+    # user must provide either a key or a where predicate
+    if where_function is not None:
+        if isinstance(obj, dict):
```

Comment:
seems like a common pattern to dispatch like this (dict/NT/dataclass) -- maybe later can extract somewhere

Reply:
Agreed, yeah. Could probably be moved near `asdict` in `my.core.common`